src/aspire/image/__init__.py

```diff
@@ -89,6 +89,29 @@ def im_filter(im, filt, *args, **kwargs):
 
     return im
 
+def im_filter_mat(im, filt):
```

Comment:
Couldn't this function simply call filt.evaluate_grid and then im_filter(im, filt)?

Reply:
It is not necessary to evaluate the CTF grid for each image. The number of CTFs is much less than the total number of images. So the im_filter_mat function is more efficient for large number of images. By the way, probably we should move those functions to preprocess.py under the utils folder.

Reply:
Maybe I'm missing something..you wouldn't do `evaluate_grid` for *each* image, but just once at the start of the function. `im` that you're passing in is multiple images anyway. Did you mean to say that you want to avoid it for each filter?

Reply:
In my case the CTF filters have been evaluated in 2D grid first then are passed to the images with the same kind of CTFs. It depends how we go through the images with the CFTs are randomly assigned. Furthermore, the grid can be defined in different way (half pixel shift or not). We need to consider it more generally during the reorganization.

Reply:
Ok..so perhaps the `im_filter` function needs to be rewritten to either accept a Filter object or a pre-evaluated filter grid of values.

I do think all `def im_*` functions should be part of the Image class - especially something as fundamental as convolving with a filter. Whether this functionality is used for pre-processing or not is irrelevant.

Reply:
Yes. I also agree that all related image operation should be defined in the image class but they can just call more general functions in preprocess.py such as downsample in different dimensions.